rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+existing permutation argument and compute the combined check of both
+permutation arguments:
+```
+Lagrange(n-ZK_ROWS) * (LookupPermutation * PermutationAggreg - 1) = 0
+```
+
+### Full list of polynomials and constraints
+
+#### Constants
+* `shift[0..14]` (`shift[0..7]` existing)
+* `zero_sigma`
+* `joint_combiner` (existing)
+* `beta` (existing)
+* `gamma` (existing)
+* `max_joint_size` (existing, integer)
+
+This results in 8 new constants.
+
+#### Polynomials without per-proof evaluations
+
+* `TableID` (verifier index)
+* `FixedTable(i)` (verifier index, existing)
+* `RuntimeTableSelector` (verifier index)
+* `SideLoadedTable(i, j)` (proof)
+
+This results in 2 new polynomials + 1 new polynomial for each column of each
+side-loaded table.
+
+#### Polynomials with per-proof evaluations + openings
+* `LookupTable` (existing)
+* `RuntimeTable(i)`
+* `FixedAndRuntimeTable`
+* `PermutationAggreg` (existing)
+* `LookupPermutation`
+* `Sigma(i)` (existing for `0 <= i < 7`, new for `7 <= i < 13`)
+
+This results in 8 new polynomial evaluations + 1 new evaluation for each column
+of the runtime table.
+
+#### Constraints
+* permutation argument (existing, remove `n-ZK_ROWS` check) (elided)
+* lookup argument (existing) (elided)
+* runtime-table consistency
+  ```
+  RuntimeTableSelector * sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+   = sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+  ```
+* fixed and runtime table evaluation
+  ```
+  FixedAndRuntimeTable
+    = sum[i=0, W, FixedTable(i) * joint_combiner^i]
+      + sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+      + TableID * joint_combiner^max_joint_size
+  ```
+* lookup permutation argument
+  ```
+  LookupPermutation
+   * PermutationAggreg
+   * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+       + gamma + beta * Sigma(7))
+   * (SideLoadedCombined(0) + gamma + beta * Sigma(8))
+   * (SideLoadedCombined(1) + gamma + beta * Sigma(9))
+   * (SideLoadedCombined(2) + gamma + beta * Sigma(10))
+   * (SideLoadedCombined(3) + gamma + beta * Sigma(11))
+   * (RuntimeTable(1) + gamma + beta * Sigma(12))
+   =
+     LookupPermutation[prev]
+     * PermutationAggreg[prev]
+     * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+         + gamma + x * beta * shift[7])
+     * (SideLoadedCombined(0) + gamma + x * beta * shift[8])
+     * (SideLoadedCombined(1) + gamma + x * beta * shift[9])
+     * (SideLoadedCombined(2) + gamma + x * beta * shift[10])
+     * (SideLoadedCombined(3) + gamma + x * beta * shift[11])
+     * (RuntimeTable(1) + gamma + x * beta * shift[12])
+  ```
+* lookup permutation initializer
+  ```
+  Lagrange(0)
+    * (LookupPermutation * (0 + gamma + beta * zero_sigma)
+      - (0 + gamma + beta * shift[len(shift)-1]))
+    = 0
+  ```
+* lookup permutation finalizer
+  ```
+  Lagrange(n-ZK_ROWS) * (LookupPermutation - 1) = 0
+  ```
+
+This results in 5 new checks, and the removal of 1 old check.
```

Comment:
Do we only pay the cost of these when specific tables are present?

Reply:
Yes, I've added a comment detailing this.